### DISCUSSION TITLE
New operation to un-merge a pull request (= reset --hard HEAD^)

### DISCUSSION THREAD

**Discussion by donquixote** [community/community#20823] [Pull Requests]
Sometimes we merge a pull request too soon.

The current way to undo is to reset the target branch in a local environment, and force-push.
Unfortunately, the pull request remains in "merged".

As an alternative, I want to propose the following:
- If the merge commit from this PR is the last commit on the target branch, provide a button "Undo merge". This will reset to the parent commit.
  - Doing this will set the PR status to "Open".
- If the merge commit from this PR is followed by other commits, provide a link to other PRs that need to be un-merged first.

**Comment by jfullstackdev** [comment]
not very sure if I'm getting it correctly, but there is the button to undo a merge pull request on GitHub Web UI

**Reply by donquixote** [reply]
I now see a "Revert" button but I think that just creates a new pull request with a revert commit.

**Comment by magikadev** [comment]
Well, with your alternative method, if we want to unmerge the commit followed by other commits, we need to do unmerge all the commits to the last commit, and then merge the unmerged commit except the commit you want to unmerge exactly.

To unmerge a pull request, we just click Revert button that creates a new PR with revert commit as you mentioned.

**Reply by donquixote** [reply]
> Well, with your alternative method, if we want to unmerge the commit followed by other commits

Something we should avoid ideally, but if we need to, then... 

> we need to do unmerge all the commits to the last commit, and then merge the unmerged commit except the commit you want to unmerge exactly.

Exactly :)
_EDIT: And each step should be initiated manually, so that the developer is fully aware of what they are doing._

> To unmerge a pull request, we just click Revert button that creates a new PR with revert commit as you mentioned.

But this leaves pollution in the history.

What I want is the equivalent of
```
git checkout --track origin/1.x
git reset --hard HEAD^
git push --force-with-lease origin 1.x
```

I don't want a return ticket, I want a "never went there" ticket :)

**Reply by magikadev** [reply]
Cool!

**Comment by dlamotte** [comment]
FWIW, I hit this too when changing the merge base of the PR. I had forgotten I had auto-merge enabled and changed the base to another branch which proceeded to auto-merge. I now have a PR stuck in merged which cannot reflect the desired state (open, unmerged). I pushed the correct details to unmerge as suggested in this post. But that doesn't change the PR.

A partial fix might be just disabling auto-merge when changing the base.
